br/pkg/lightning/duplicate/worker.go

```diff
@@ -0,0 +1,203 @@
+	"bytes"
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/br/pkg/lightning/log"
+	"github.com/pingcap/tidb/util/extsort"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+)
+
+type task struct {
+	startKey internalKey
+	endKey   internalKey
+}
+
+type worker struct {
+	sorter  extsort.ExternalSorter
+	taskCh  chan task
+	taskWg  *sync.WaitGroup
+	numDups *atomic.Int64
+	handler Handler
+	logger  log.Logger
+}
+
+func (w *worker) run(ctx context.Context) error {
+	for {
+		select {
+		case t, ok := <-w.taskCh:
+			if !ok {
+				return nil
+			}
+			if err := w.runTask(ctx, t); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (w *worker) runTask(ctx context.Context, t task) (retErr error) {
+	logTask := w.logger.With(
+		zap.Stringer("startKey", t.startKey),
+		zap.Stringer("initialEndKey", t.endKey),
+	).Begin(zap.InfoLevel, "[dup-detector] runTask")
+
+	var processedKeys int64
+
+	defer func() {
+		w.taskWg.Done()
+		err := w.handler.Close()
+		if retErr == nil {
+			retErr = err
+		}
+		logTask.End(zap.ErrorLevel, retErr,
+			zap.Stringer("endKey", t.endKey),
+			zap.Int64("processedKeys", processedKeys),
+		)
+	}()
+
+	it, err := w.sorter.NewIterator(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = it.Close()
+	}()
+
+	// Periodically check context cancellation or task splitting during iteration.
+	const checkInterval = 1000
+	numIterations := 0
+
+	inDup := false
+	var prevKey, curKey internalKey
+	for it.Seek(encodeInternalKey(nil, t.startKey)); it.Valid(); it.Next() {
+		if err := decodeInternalKey(it.UnsafeKey(), &curKey); err != nil {
+			return err
+		}
+		if compareInternalKey(curKey, t.endKey) >= 0 {
+			break
+		}
+		processedKeys++
+
+		if bytes.Equal(curKey.key, prevKey.key) {
+			if inDup {
+				if err := w.handler.Append(curKey.keyID); err != nil {
+					return err
+				}
+			} else {
+				if err := w.handler.Begin(curKey.key); err != nil {
+					return err
+				}
+				if err := w.handler.Append(prevKey.keyID); err != nil {
+					return err
+				}
+				if err := w.handler.Append(curKey.keyID); err != nil {
+					return err
+				}
+				inDup = true
+				w.numDups.Add(1)
+			}
+		} else if inDup {
+			if err := w.handler.End(); err != nil {
+				return err
+			}
+			inDup = false
+		}
+
+		numIterations++
+		// Try to split the task and send the new task to the task channel
+		// so that other idle workers can process it.
+		if numIterations%checkInterval == 0 {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
+			if len(w.taskCh) == 0 {
+				userSplitKey := genSplitKey(curKey.key, t.endKey.key)
+				if bytes.Compare(userSplitKey, curKey.key) > 0 {
+					splitKey := internalKey{key: userSplitKey}
+					w.taskWg.Add(1)
+					select {
+					case w.taskCh <- task{startKey: splitKey, endKey: t.endKey}:
+						t.endKey = splitKey
+					default:
+						w.taskWg.Done()
+					}
+				}
+			}
+		}
+		prevKey, curKey = curKey, prevKey
```

Comment:
Should be prev, cur = cur, next? I'm a bit dizzy today 😴

Reply:
`prevKey` and `curKey` are reused as buffers to reduce allocation. `curKey` in this iteration will be `prevKey` in the next iteration, and the `prevKey` in this iteration will be reused as curKey.

Reply:
OK, please left a comment because it's a bit hard to understand